src/java.base/share/classes/com/sun/crypto/provider/HKDFKeyDerivation.java

```diff
@@ -0,0 +1,414 @@
+     *         {@code null} if no info data is provided.
+     * @param outLen
+     *         the length in bytes of the required output
+     *
+     * @return a byte array containing the complete {@code KDF} output.  This
+     *         will be at least as long as the requested length in the
+     *         {@code outLen} parameter, but will be rounded up to the nearest
+     *         multiple of the HMAC output length.
+     *
+     * @throws InvalidKeyException
+     *         if an invalid PRK was provided through the
+     *         {@code HKDFParameterSpec} or derived during the extract phase.
+     */
+    private byte[] hkdfExpand(byte[] prk, byte[] info, int outLen)
+            throws InvalidKeyException, NoSuchAlgorithmException {
+        byte[] kdfOutput;
+
+        if (prk == null || prk.length < hmacLen) {
+            throw new InvalidKeyException(
+                    "prk must be at least " + hmacLen + " bytes");
+        }
+
+        SecretKey pseudoRandomKey = new SecretKeySpec(prk, hmacAlgName);
+
+        Mac hmacObj = Mac.getInstance(hmacAlgName);
+
+        // Calculate the number of rounds of HMAC that are needed to
+        // meet the requested data.  Then set up the buffers we will need.
+        hmacObj.init(pseudoRandomKey);
+        int rounds = (outLen + hmacLen - 1) / hmacLen;
+        kdfOutput = new byte[outLen];
+        int i = 0;
+        int offset = 0;
+        try {
+            while (i < rounds) {
+                if (i > 0) {
+                    hmacObj.update(kdfOutput, offset - hmacLen,
+                                   hmacLen); // add T(i-1)
+                }
+                hmacObj.update(info);                   // Add info
+                hmacObj.update((byte) ++i);             // Add round number
+                if (i == rounds && (outLen - offset < hmacLen)) {
+                    // special handling for last chunk
+                    byte[] tmp = hmacObj.doFinal();
+                    System.arraycopy(tmp, 0, kdfOutput, offset,
+                                     outLen - offset);
+                    Arrays.fill(tmp, (byte) 0x00);
+                    offset = outLen;
+                } else {
+                    hmacObj.doFinal(kdfOutput, offset);
+                    offset += hmacLen;
+                }
+            }
+        } catch (ShortBufferException sbe) {
+            // This really shouldn't happen given that we've
+            // sized the buffers to their largest possible size up-front,
+            // but just in case...
+            throw new ProviderException(sbe);
+        }
+        return kdfOutput;
+    }
+
+    protected KDFParameters engineGetParameters() {
+        return null;
+    }
+
+    public static final class HKDFSHA256 extends HKDFKeyDerivation {
+        public HKDFSHA256(KDFParameters kdfParameters)
+                throws InvalidAlgorithmParameterException {
+            super("HmacSHA256", SHA256_HMAC_SIZE, kdfParameters);
```

Comment:
Using the enum, this line would be:
`            super(SupportedHmac.SHA256, kdfParameters);`

Reply:
Similar changes to the other child classes below.